docs/proposals/role-entry-points.md

```diff
@@ -0,0 +1,264 @@
+# Entry-point based roles
+
+## We have a problem
+
+Each role consists of a set of tasks and variables read/set in some of the tasks.
+Some of the tasks can store their results into registered variables,
+e.g. store stdout of a docker pull command.
+Some of the tasks may be run only under certain condition that are generated
+outside of a role, e.g. pull images only if a deployment is containerized.
+Some of the tasks are run only if a certain variable is defined,
+e.g. add additional Kubelet arguments into the `node-config.yml`
+if `kubelet_additional_args` inventory variable is defined.
+
+Some roles work with a lot of variables that from a point of a role are uninvited citizens.
+Some of the variables are set inside a role, some of them are set outside of a role.
+The variables outside of a role are hard to track, hard to determine their value or
+a condition under which the variables are defined. These unknown facts make a role harder
+to maintain, harder to make a role closed with respect to variables and to make
+a role practically modular.
+
+## Proposal
+
+Make a role that conforms to the following conditions:
+* none of the variables read/set inside a role are global
+* all variables inside a role are `r_` (`r` standing for a role) and/or `l_`(`l` standing for a local) prefixed
```

Comment:
I have never really like l_ variables but I do like how they tell me that vars have been created locally.  It takes me time to process the l because it looks like any of the following chars, 1, l, I, i, L, | and I think it's generally bad practice to call variables by those names.

Could we just use the r_ prefix and still get the same benefit?  This would denote that the variable was passed in, created within the role, or is a role default.  At least this would standardize on variables.